components/director/internal/domain/formation/service.go

```diff
@@ -889,64 +881,140 @@ func (s *service) UnassignFormation(ctx context.Context, tnt, objectID string, o
+	// So if the async notification was sent, and we're processing the sync notification, meanwhile the async participant sends
+	// FA status update request, he will have the latest state of the formation assignment even when we didn't finish the sync notification processing.
+	if err := s.executeInTransaction(ctx, func(ctxWithTransact context.Context) error {
+		for _, ia := range initialAssignmentsData {
+			log.C(ctx).Infof("Update and persist in the DB '%s' state of formation assignment with ID: '%s'", ia.State, ia.ID)
+			faClone := ia.Clone()
+			faClone.State = string(model.DeletingAssignmentState)
+			formationassignment.ResetAssignmentConfigAndError(faClone)
+			if err := s.formationAssignmentService.Update(ctxWithTransact, faClone.ID, faClone); err != nil {
+				return errors.Wrapf(err, "while updating formation assignment with ID: '%s' to '%s' state", ia.ID, ia.State)
+			}
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err == nil {
+			return
+		}
+
+		if !shouldRevertAssignmentsUpdateFromTransaction {
+			return
+		}
+
+		log.C(ctx).Infof("Reverting formation assignment changes that updated them to DELETING state and reset their configuration in the first transaction...")
+		if terr := s.executeInTransaction(ctx, func(ctxWithTransact context.Context) error {
+			for _, fa := range initialAssignmentsData {
+				if updateErr := s.formationAssignmentService.Update(ctxWithTransact, fa.ID, fa); updateErr != nil {
+					log.C(ctx).WithError(updateErr).Errorf("while updating formation assignment with ID: %s", fa.ID)
+					return updateErr
+				}
+			}
+
+			return nil
+		}); terr != nil {
+			log.C(ctx).WithError(terr).Error("An error occurred while reverting formation assignments with their original data")
+		}
+	}()
+
 	// It is important that all operations regarding formation assignments should be in the inner transaction
-	tx, err := s.transact.Begin()
+	tx, terr := s.transact.Begin()
+	err = terr
 	if err != nil {
 		return nil, err
 	}
 	transactionCtx := persistence.SaveToContext(ctx, tx)
 	defer s.transact.RollbackUnlessCommitted(transactionCtx, tx)
 
-	requests, err := s.notificationsService.GenerateFormationAssignmentNotifications(transactionCtx, tnt, objectID, formationFromDB, model.UnassignFormation, objectType)
+	requests, nerr := s.notificationsService.GenerateFormationAssignmentNotifications(transactionCtx, tnt, objectID, formationFromDB, model.UnassignFormation, objectType)
+	err = nerr
 	if err != nil {
 		if apperrors.IsNotFoundError(err) {
-			commitErr := tx.Commit()
-			if commitErr != nil {
-				return nil, errors.Wrapf(err, "while committing transaction with error")
+			if commitErr := tx.Commit(); commitErr != nil {
+				err = errors.Wrapf(
+					errors.Wrapf(commitErr, "while committing transaction"),
+					errors.Wrapf(nerr, "while generating formation assignment notifications").Error(),
```

Comment:
`err` already contains the data from `nerr` so line 956 is not needed it will only duplicate the message inside. Also do we really need the commit error itself. Isnt the message "while committing transaction with error" enough to show what happend?

Reply:
We have two operations here that's why I wanted to log both of them so we could have better log and it will be easier to know the transaction error occurred while another error happened(generate notifications).
Also, the log shouldn't be duplicated? It should be overridden. The initial value of err will be what we have in `nerr` but to 954 line we assign another value.